lib/luaxtable.c

```diff
@@ -46,34 +51,154 @@ typedef struct luaxtable_s {
+static inline void luaxtable_newdata(lua_State *L, void *ptr, size_t size, bool sleep)
 {
-	return 0;
+	lunatik_require(L, data);
+	lunatik_object_t *data = lunatik_checknull(L, luadata_new(ptr, size, sleep));
+	lunatik_cloneobject(L, data);
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+#define luaxtable_getrcu(xtable, hook, ttype, par)			\
+do {									\
+	const char *name = par->hook->name;				\
+	uint8_t revision = par->hook->revision;				\
+	unsigned short family = par->hook->family;			\
+									\
+	char *key = NULL;						\
+	int keylen = luaxtable_key(&key, name, revision, family, ttype);	\
+	if (keylen < 0) {						\
+		pr_err("unable to allocate key\n");			\
+		return -1;						\
+	}								\
+									\
+	lunatik_object_t *obj = luarcu_gettable(xtable_##hook##hooks, key, keylen);	\
+	if (obj == NULL) {						\
+		lunatik_free(key);					\
+		pr_err("could not find table for %s\n", name);		\
+		return -1;						\
+	}								\
+	lunatik_free(key);						\
+	xtable = (luaxtable_t *)obj->private;				\
+} while(0)
+
+#define luaxtable_setrcu(object, hook, HOOK)					\
+do {									\
+	char *key = NULL;						\
+	int  keylen = luaxtable_key(&key, hook->name, hook->revision, hook->family, HOOK);	\
+	if (keylen < 0) {						\
+		luaL_error(L, "unable to allocate key");		\
+		return -1;						\
+	}								\
+									\
+	if (luarcu_settable(xtable_##hook##hooks, key, keylen, object) != 0) {	\
+		lunatik_free(key);					\
+		luaL_error(L, "unable to set table");			\
+		return -1;						\
+	}								\
+	lunatik_free(key);						\
+} while(0)
+
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, void *skb, int skb_size, void *par, int par_size, const char *op, int nargs, int nret)
 {
+	int base = lua_gettop(L);
+	int ret = -ENXIO;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("could not find ops table\n");
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	luaxtable_newdata(L, par, par_size, false);
+	if (nargs == 2 )
+		luaxtable_newdata(L, skb, skb_size, false);
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		ret = -ECANCELED;
+		goto err;
+	}
+
+	if (nret > 0)
+		ret = luaL_optinteger(L, -1, 0);
+	else
+		ret = 0;
+
+	return ret;
+err:
+	lua_settop(L, base); /* pop everything, including args */
+	return ret;
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
-{
-	return NF_ACCEPT;
+#define LUAXTABLE_HOOK_CB(T, U, V, hook, huk) 				\
+static T luaxtable_##hook(U skb, V par)					\
+{									\
+	int ret = NF_DROP;						\
+	const luaxtable_info_t *info = (const luaxtable_info_t *)par->huk##info;	\
+	luaxtable_t *xtable = info->data;				\
+									\
+	lunatik_run(xtable->runtime, luaxtable_invoke, ret, xtable, (void *)skb, sizeof(struct sk_buff), (void *)par, sizeof(struct xt_action_param), #hook, 2, 1);					\
```

Comment:
why do you need the sizeofs here? and to use `void *`? if you have different parameters for target and match, let's just have to separate "invoke" functions..

Reply:
The following are the struct parameters that are used by match and target functions : 

- `match` : struct sk_buff, **const** struct **xt_action_param**
- `target` : **const** struct sk_buff, struct **xt_action_param**

- `match_checkentry` : const struct xt_**mtchk**_param
- `target_checkentry` : const struct xt_**tgchk**_param

- `match_destroy` : const struct xt_**mtdtor**_param
- `target_destory` : const struct xt_**tgdtor**_param

There are 5 different structs for us to handle.
These have a common pattern (just match is replaced by target, for the fields).

I used `void *` and `sizeof` for luadata object allocation.